components/ws-proxy/pkg/proxy/infoprovider.go

```diff
@@ -351,15 +379,48 @@ func (c *workspaceInfoCache) delete(workspaceID string) {
 	delete(c.infos, workspaceID)
 }
 
-func (c *workspaceInfoCache) getByID(workspaceID string) (*WorkspaceInfo, bool) {
+func (c *workspaceInfoCache) WaitFor(ctx context.Context, workspaceID string) (w *WorkspaceInfo, ok bool) {
 	c.mu.RLock()
-	defer c.mu.RUnlock()
+	w, ok = c.infos[workspaceID]
+	c.mu.RUnlock()
+	if ok {
+		return
+	}
+
+	inc := make(chan *WorkspaceInfo)
+	go func() {
```

Comment:
Do I understand correctly that we need this goroutine because we want to wait for - `select` - on a condition variable which is not `select`able? And we use `cond` because we don't have the buffering problems and the `cond` code is nicer in the other places?

Reply:
Exactly, we have two things we want to wait for: context cancelation or info becoming available. We cannot select on the cond, hence need to "turn it into a channel". We use a `cond` because we have potentially multiple `WaitFor`s waiting, and we cannot use a single channel directly (except with a much more complicated locked map).